src/main/java/org/folio/service/transactions/batch/BatchPaymentCreditService.java

```diff
@@ -0,0 +1,168 @@
+package org.folio.service.transactions.batch;
+
+import org.folio.rest.jaxrs.model.Budget;
+import org.folio.rest.jaxrs.model.Encumbrance;
+import org.folio.rest.jaxrs.model.Transaction;
+import org.folio.rest.jaxrs.model.Transaction.TransactionType;
+import org.folio.utils.MoneyUtils;
+import org.javamoney.moneta.Money;
+import org.javamoney.moneta.function.MonetaryFunctions;
+
+import javax.money.CurrencyUnit;
+import javax.money.Monetary;
+import javax.money.MonetaryAmount;
+import java.util.List;
+import java.util.Map;
+
+import static org.folio.rest.jaxrs.model.Transaction.TransactionType.PAYMENT;
+import static org.folio.utils.CalculationUtils.calculateBudgetSummaryFields;
+import static org.folio.utils.MoneyUtils.subtractMoney;
+import static org.folio.utils.MoneyUtils.sumMoney;
+
+public class BatchPaymentCreditService extends AbstractBatchTransactionService {
+
+  @Override
+  public TransactionType getTransactionType() {
+    return PAYMENT;
+  }
+
+  @Override
+  public void prepareCreatingTransactions(List<Transaction> transactionsToCreate, BatchTransactionHolder holder) {
+    updateEncumbrances(transactionsToCreate, holder);
+    updateBudgets(transactionsToCreate, holder);
+    markPendingPaymentsForDeletion(holder);
+  }
+
+  @Override
+  public void prepareUpdatingTransactions(List<Transaction> transactionsToUpdate, BatchTransactionHolder holder) {
+    updateEncumbrances(transactionsToUpdate, holder);
+    updateBudgets(transactionsToUpdate, holder);
+  }
+
+  private void updateEncumbrances(List<Transaction> transactions, BatchTransactionHolder holder) {
+    List<Transaction> encumbrances = prepareEncumbrancesToProcess(transactions, holder, Transaction::getPaymentEncumbranceId);
+    if (encumbrances.isEmpty()) {
+      return;
+    }
+    Map<String, Transaction> encumbranceMap = holder.getLinkedEncumbranceMap();
+    Map<String, Transaction> existingTransactionMap = holder.getExistingTransactionMap();
+    List<Transaction> transactionsWithAnEncumbrance = transactions.stream()
+      .filter(tr -> tr.getPaymentEncumbranceId() != null)
```

Comment:
Better to use Objects::nonNull lambda

Reply:
This is not filtering for non-null transactions here. I don't know how I could use Objects::nonNull in this case...

Reply:
It's just me, sorry.